### PR TITLE
Add FastAPI support with an ASGI Middleware

### PR DESCRIPTION
**Pull request by artefactop**
Can be easily extensible for other ASGI frameworks.

### CODE REVIEW

**Comment by sjs994** [comment]
Thanks @artefactop for the PR. Will take a look.

File: python/sqlcommenter-python/google/cloud/sqlcommenter/fastapi.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/python
```
Comment (sjs994) [inline python/sqlcommenter-python/google/cloud/sqlcommenter/fastapi.py:1]: Although SQLCommenter is donated to OpenTelemetry, but the migration is going to take time. For now, should we add the copyright as added in all other existing files in this repo's python packages ?

File: python/sqlcommenter-python/google/cloud/sqlcommenter/fastapi.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+class SQLCommenterMiddleware:
+    """The ASGI application middleware.
+    This class is an ASGI middleware that augment SQL statements before execution,
+     with comments containing information about the code that caused its execution.
+
+    Args:
+        app: The ASGI application callable to forward requests to.
+    """
+
+    def __init__(
```
Comment (sjs994) [inline python/sqlcommenter-python/google/cloud/sqlcommenter/fastapi.py:39]: nit: formatting
```
def __init__(self, app):
```

File: python/sqlcommenter-python/google/cloud/sqlcommenter/fastapi.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            self,
+            app,
+    ):
+        self.app = guarantee_single_callable(app)
+
+    async def __call__(self, scope, receive, send):
+        """The ASGI application
+        Args:
+            scope: A ASGI environment.
```
Comment (sjs994) [inline python/sqlcommenter-python/google/cloud/sqlcommenter/fastapi.py:48]: An ?

File: python/sqlcommenter-python/google/cloud/sqlcommenter/sqlalchemy/executor.py [hunk 4]
```diff
@@ -54,7 +55,10 @@ def before_cursor_execute(conn, cursor, sql, parameters, context, executemany):
         # folks using it in a web framework such as flask will
         # use it in unison with flask but initialize the parts disjointly,
         # unlike Django which uses ORMs directly as part of the framework.
-        data.update(get_flask_info())
+        info = get_flask_info()
+        if not info:
+            info = get_fastapi_info()
+        data.update(info)
```
Comment (sjs994) [inline python/sqlcommenter-python/google/cloud/sqlcommenter/sqlalchemy/executor.py:61]: I am a little concerned on the scalability of this solution. If we add more frameworks, there will be more if and elseif statements added. Is it possible to think of an approach to use a switch-case or map based solution something like:

```
switch (get_framework()) {
  case "FastAPI":
     return get_fast_api_info()
  case "Flask":
    return get_flask_info()
  default:
   return None
}
```

File: python/sqlcommenter-python/tests/fastapi/tests.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+@pytest.fixture
+def client():
+    client = TestClient(app)
+    yield client
+
+
+def test_get_fastapi_info_in_request_context(client):
+    expected = {
+        'app_name': 'FastAPI',
```
Comment (sjs994) [inline python/sqlcommenter-python/tests/fastapi/tests.py:36]: Is the default application name "FastAPI" or can we set this to something else ?
Just seems confusing as the `app name == framework name`

**Review by sjs994 — COMMENTED** [review]
Thanks @artefactop for the change. Apart from few comments added, looks good to me.

File: python/sqlcommenter-python/google/cloud/sqlcommenter/sqlalchemy/executor.py (hunk 4 above)
Comment (artefactop) [reply python/sqlcommenter-python/google/cloud/sqlcommenter/sqlalchemy/executor.py:61]: The only way we can do something like that is adding which framework are we using in `BeforeExecuteFactory`, as it should just work without configuration I don't know if that will be the better choice

File: python/sqlcommenter-python/tests/fastapi/tests.py (hunk 5 above)
Comment (artefactop) [reply python/sqlcommenter-python/tests/fastapi/tests.py]: I changed by SQLCommenter

File: python/sqlcommenter-python/google/cloud/sqlcommenter/sqlalchemy/executor.py (hunk 4 above)
Comment (sjs994) [reply python/sqlcommenter-python/google/cloud/sqlcommenter/sqlalchemy/executor.py]: Thanks, refactoring into separate method looks better

**Review by sjs994 — APPROVED** [review]

File: python/sqlcommenter-python/tests/sqlalchemy/tests.py [hunk 6]
```diff
@@ -126,3 +126,38 @@ def test_route_disabled(self, get_info):
...
+    fastapi_info = {
+        'framework': 'fastapi',
+        'controller': 'c',
+        'route': '/',
+    }
+
+    @mock.patch('google.cloud.sqlcommenter.sqlalchemy.executor.get_fastapi_info', return_value=fastapi_info)
+    def test_all_data(self, get_info):
+        self.assertSQL(
+            "SELECT 1; /*controller='c',framework='fastapi',route='/'*/",
```
Comment (sjs994) [inline python/sqlcommenter-python/tests/sqlalchemy/tests.py:141]: Sorry to miss this one. 
Although, this change looks good, wanted to give you an FYI on an existing bug in SQLAlchemy and psycopg2 sqlcommenter:

This query:
```
SELECT 1; /*controller='c',framework='fastapi',route='/'*/
```

In general a `;` ends a SQL query. So, any comment added after that is not really part of that query. So, even with this change, my guess is SQL comments would not work as expected as DB will assume there are 2 queries:
1. SELECT 1;
2. /\*controller='c',framework='fastapi',route='/'\*/

I have added one small [PR](https://github.com/google/sqlcommenter/pull/112) for psycopg2 as of now (Waiting for review) and will create another for SQLAlchemy if the approach doesn't have major concern.

**Comment by artefactop** [comment]
@sjs994 should I merge this PR or you do?

**Comment by sjs994** [comment]
@artefactop I can do that.

**Pull request merged**